### PR TITLE
Fix abstract replay buffer interface

### PR DESCRIPTION
**Pull request by maliesa96**
Leaving this open for us to discuss the improved abstract replay buffer interface. I've kept the initial API as simple as possible. As is, `PathBuffer` can be easily modified to subclass `ReplayBuffer`, with the caveat that if an algorithm calls `ReplayBuffer.add_transitions()`, future calls to `sample_path` would no longer be valid, which I think makes this a little hacky. On the other hand, we could scrap `add_transitions` altogether.

### CODE REVIEW

File: src/garage/replay_buffer/replay_buffer.py [hunk 1]
```diff
@@ -12,173 +12,75 @@
...
+        Raises:
+            NotYetImplementedError: If the replay buffer does not support
+                path sampling.
         """
-        transition = {k: [v] for k, v in kwargs.items()}
-        self.add_transitions(**transition)
-
-    def add_transitions(self, **kwargs):
-        """Add multiple transitions into the replay buffer.
+        raise NotImplementedError
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:54]: ABCs don't need `NotImplementedError` or `pass` -- just leave them blank

File: src/garage/replay_buffer/replay_buffer.py [hunk 2]
```diff
@@ -12,173 +12,75 @@
...
+            batch_size (int): The number of transitions to be sampled.
 
+        Returns:
+            SampleBatch: batch_size transitions sampled.
         """
-        raise NotImplementedError
 
-    def add_transition(self, **kwargs):
-        """Add one transition into the replay buffer.
+    def sample_path(self):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:44]: mark with `@abc.abstractmethod`

File: src/garage/replay_buffer/replay_buffer.py [hunk 3]
```diff
@@ -12,173 +12,75 @@
 import abc
 from abc import abstractmethod
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:13]: please use only package-level imports for non-garage packages (i.e. use this as `@abc.abstractmethod`)

File: src/garage/replay_buffer/replay_buffer.py [hunk 4]
```diff
@@ -12,173 +12,75 @@
...
 
 class ReplayBuffer(metaclass=abc.ABCMeta):
     """Abstract class for Replay Buffer.
 
     Args:
         env_spec (garage.envs.EnvSpec): Environment specification.
-        size_in_transitions (int): total size of transitions in the buffer
-        time_horizon (int): time horizon of rollout.
-
+        capacity_in_transitions (int): capacity of the replay buffer,
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:21]: we should keep this terminology consistent in the interfaces

do we have Transitions, TimeSteps, Steps, or Samples? 

I slightly prefer TimeSteps and I think we already have that in dtypes, so this would be `capacity_in_time_steps`. And `SampleBatch` should actually be `TimeStepBatch`

If we thinks that's too wordy, we should look at the entire termnology family.

File: src/garage/replay_buffer/replay_buffer.py [hunk 5]
```diff
@@ -12,173 +12,75 @@
...
+            measured in transitions.
     """
 
-    def __init__(self, env_spec, size_in_transitions, time_horizon):
-        del env_spec
+    def __init__(self, env_spec, capacity_in_transitions):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:25]: I think that interfaces should be constructor-less and have no default implementations. If there's an obvious typical implementation that can be a simple abstract class inherited from the interface.

For instance, there's no reason a replay buffer doesn't need to know the capacity up front, in which case forcing this constructor args is just a useless burden.

File: src/garage/replay_buffer/replay_buffer.py [hunk 6]
```diff
@@ -12,173 +12,75 @@
...
-        for key in self._buffer:
-            self._buffer[key][idx] = episode_buffer[key]
-        self._n_transitions_stored = min(
-            self._size_in_transitions, self._n_transitions_stored +
-            self._time_horizon * rollout_batch_size)
 
     @abstractmethod
-    def sample(self, batch_size):
-        """Sample a transition of batch_size.
+    def sample_transitions(self, batch_size):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:34]: `sample_time_step_batch` or simply `sample_batch`

File: src/garage/replay_buffer/replay_buffer.py [hunk 7]
```diff
@@ -12,173 +12,75 @@
...
-        """Add multiple transitions into the replay buffer.
+        raise NotImplementedError
 
-        A transition contains one or multiple entries, e.g.
-        observation, action, reward, terminal and next_observation.
-        The same entry of all the transitions are stacked, e.g.
-        {'observation': [obs1, obs2, obs3]} where obs1 is one
-        numpy.ndarray observation from the environment.
+    @abstractmethod
+    def add_path(self, path):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:57]: `add_trajectory` since the dtype is `TrajectoryBatch`

consider defining a dtype `Trajectory` if the singleton case appears often.

File: src/garage/replay_buffer/replay_buffer.py [hunk 8]
```diff
@@ -12,173 +12,75 @@
...
-    def _convert_episode_to_batch_major(self):
-        """Convert the shape of episode_buffer.
-
-        episode_buffer: {time_horizon, algo.rollout_batch_size, flat_dim}.
-        buffer: {size, time_horizon, flat_dim}.
-
-        Returns:
-            dict: Transitions that have been formated to fit properly in this
-                replay buffer.
+            transitions (dict[str, np.ndarray] or TrajectoryBatch): transitions to be added to the buffer.
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:69]: time_steps

File: src/garage/replay_buffer/replay_buffer.py [hunk 9]
```diff
@@ -12,173 +12,75 @@
...
-            values = np.array(value)
-            self._buffer[key] = np.zeros(
-                [self._size, self._time_horizon, *values.shape[1:]],
-                dtype=values.dtype)
-        self._initialized_buffer = True
-
-    def _get_storage_idx(self, size_increment=1):
-        """Get the storage index for the episode to add into the buffer.
+    @abstractmethod
+    def add_transitions(self, transitions):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:65]: `add_time_step_batch`

Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:65]: i think this should accept only `TimeStepBatch` or possible `Union[TimeStep, TimeStepBatch]`

File: src/garage/replay_buffer/replay_buffer.py (hunk 7 above)
Comment (ryanjulian) [reply src/garage/replay_buffer/replay_buffer.py:57]: if `Trajectory` dtype exists, this could accept `Union[Trajectory, TrajectoryBatch]`

File: src/garage/replay_buffer/replay_buffer.py [hunk 10]
```diff
@@ -12,173 +12,75 @@
...
     def full(self):
-        """Whether the buffer is full.
-
-        Returns:
-            bool: True of the buffer has reachd its maximum size.
-                False otherwise.
-
-        """
-        return self._current_size == self._size
+        """bool: True if the buffer has reached its maximum capacity."""
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:75]: Hmm does this *really* need to be in the interface? what is it used for?

File: src/garage/replay_buffer/replay_buffer.py [hunk 11]
```diff
@@ -12,173 +12,75 @@
 import abc
 from abc import abstractmethod
 
-import numpy as np
-
 
 class ReplayBuffer(metaclass=abc.ABCMeta):
```
Comment (ryanjulian) [inline src/garage/replay_buffer/replay_buffer.py:16]: Immediately I see an issue here where there is 2 interfaces: a `TimeStep`-oriented one and `Trajectory`-oriented one.

Do we want to:
* Force all replay buffers to implement both semantics?
* Standardize on adding/sampling time steps?
* Standardize on adding/sampling trajectories? Note this would make single-stepping sampling a pain.
* Choose one interface as the universal case, define a more-specialized interface for the less-universal case? In this case, I think `TimeStep`-oriented is the universal case and `Trajectory`-oriented is the specific case.
* Define two different interfaces, entirely.

File: src/garage/replay_buffer/replay_buffer.py (hunk 5 above)
Comment (krzentner) [reply src/garage/replay_buffer/replay_buffer.py:25]: My personal opinion on this topic is very simple: `__init__` should not be part of an API, ever. When I'm constructing something, I know exactly what type it is, so the API doesn't mean anything. The only reason to have `__init__` be part of an API is that you want to have default method implementations in terms of private fields, which I think is an anti-pattern.

We have some places where we've done this already in garage, but it's resulted in difficult to explain or fix bugs (see torch policies / issue #1141). The problem is that implementing multiple interfaces on the same type (which is very common in python, once you start looking) then requires reasoning about multiple inheritance, which will always be somewhat broken.

**Comment by ryanjulian** [comment]
@krzentner since the goal of this PR is to establish a durable ReplayBuffer API, can you give us some of your thoughts on what you think that should look like?

I agree the data backend API should be as simple as possible. We could go so far as making it agnostic of the stored datatype (doesn't care if it's storing trajectories or samples).

**Comment by maliesa96** [comment]
@ryanjulian Regarding the issue you point out, the current interface assumes we're going with option 1 (force replay buffers to implement both semantics), mainly because it was simpler to get this PR up.
I think the cleaner approach is to have separate APIs for TimeStep and Trajectory based replay buffers. This removes one layer of abstraction, but it makes reading code easier because I'd be able to tell whether an algorithm uses one type of buffer or the other simply based on the calls it makes. It also makes implementations of the buffers more straightforward, they either deal with TrajectoryBatch or TimeSteps, not both. This way we can also keep the simpler PathBuffer implementation as is like @krzentner mentioned.

**Comment by ryanjulian** [comment]
Perhaps we should provide a `ReplayBuffer` interface which is data-agnostic, and further `PathBuffer` and `SampleBuffer` interfaces which specify the data types which are being stored.

```python
class ReplayMemory(abc.ABC):

    @abc.abstractmethod
    def puts(self, samples: Sequence[Any]): -> Sequence[int]

    @abc.abstractmethod
    def gets(self, indices): -> Sequence[Any]

    @abc.abtractemthod
    def evicts(self, indices): --> Sequence[int]

    @property
    @abc.abstractmethod
    def size(self): --> int

    @property
    @abc.abstractmethod
    def max_index(self): --> int

    @property
    @abc.abstractmethod
    def min_index(self): --> int


class ReplayBuffer:

    @abc.abstractmethod
    def puts(self, samples: Sequence[Any]): -> Sequence[int]

    @abc.abstractmethod
    def sample(self, n: int): -> Sequence[Any]
    

class FIFOReplayBuffer(ReplayBuffer):

    def __init__(self, memory: ReplayMemory, capacity_in_samples: int):
        self._capacity = capacity_in_samples
        self._memory = memory

    def puts(self, samples: Sequence[Any]): -> Sequence[int]
        maybe_overflow = self._memory.size + len(samples) - self._capacity
        if maybe_overflow > 0:
            self._memory.evicts(list(range(min_index, min_index + maybe_overflow)))

        return self._memory.puts(samples)

    def sample(self, n: int): -> Sequence[Any]
        indices = random.sample(range(self._memory.min_index, self._memory.max_index), n)
        return self._memory.gets(indices)
    

def FIFOPathBuffer(FIFOReplayBuffer):

    def puts(self, samples: TrajectoryBatch): -> Sequence[int]
        return super().puts(samples.to_list())

    def sample(self, n: int): -> TrajectoryBatch
        trajs = super().sample(n)
        return TrajectoryBatch.from_list(trajs)


def FIFOTimeStepBuffer(FIFOReplayBuffer):

    def puts(self, samples: TimeStepBatch): -> Sequence[int]
        return super().puts(samples.to_list())

    def sample(self, n: int): -> TimeStepBatch
        ts = super().sample(n)
        return TimeStepBatch.from_list(ts)
